examples/cartridge_self_study/synthesize.py

```diff
@@ -0,0 +1,199 @@
+    num_samples: int,
+    seed_prompt_types: list[str],
+    min_tokens_per_chunk: int,
+    max_tokens_per_chunk: int,
+    max_new_tokens: int,
+    temperature: float,
+    top_p: float,
+):
+    output_path.parent.mkdir(parents=True, exist_ok=True)
+    if output_path.exists():
+        output_path.unlink()
+
+    for t in seed_prompt_types:
+        if t not in DEFAULT_SEED_PROMPTS:
+            raise ValueError(f"Unknown seed prompt type '{t}', expected one of: {sorted(DEFAULT_SEED_PROMPTS)}")
+
+    corpus_token_ids = tokenizer(
+        corpus_text,
+        return_tensors=None,
+        add_special_tokens=False,
+        truncation=False,
+    )["input_ids"]
+    chunks = _chunk_token_ids(corpus_token_ids, min_tokens=min_tokens_per_chunk, max_tokens=max_tokens_per_chunk)
+    if not chunks:
+        raise ValueError("Corpus too small after chunking; try lowering `min_tokens_per_chunk`.")
+
+    device = getattr(model, "device", torch.device("cpu"))
+    model.eval()
+
+    rng = torch.Generator(device="cpu")
+    for _ in range(num_samples):
+        chunk_idx = int(torch.randint(low=0, high=len(chunks), size=(1,), generator=rng).item())
+        prompt_idx = int(torch.randint(low=0, high=len(seed_prompt_types), size=(1,), generator=rng).item())
+
+        chunk_text = tokenizer.decode(chunks[chunk_idx], skip_special_tokens=True)
+        seed_prompt = DEFAULT_SEED_PROMPTS[seed_prompt_types[prompt_idx]]
+
+        teacher_prompt = _apply_chat_template(
+            tokenizer,
+            [{"role": "system", "content": chunk_text}, {"role": "user", "content": seed_prompt}],
+            add_generation_prompt=True,
+        )
+        student_prompt = _apply_chat_template(
+            tokenizer,
+            [{"role": "user", "content": seed_prompt}],
+            add_generation_prompt=True,
+        )
+        if teacher_prompt is None or student_prompt is None:
+            ctx_ids = tokenizer(chunk_text, add_special_tokens=False)["input_ids"]
+            x_ids = tokenizer(seed_prompt, add_special_tokens=False)["input_ids"]
+            teacher_prompt = torch.tensor(ctx_ids + x_ids, dtype=torch.long).unsqueeze(0)
+            student_prompt = torch.tensor(x_ids, dtype=torch.long).unsqueeze(0)
+
+        teacher_prompt = teacher_prompt.to(device)
+        student_prompt = student_prompt.to(device)
+
+        gen_kwargs = {
+            "max_new_tokens": max_new_tokens,
+            "do_sample": temperature > 0,
+            "temperature": max(temperature, 1e-5),
+            "top_p": top_p,
+            "pad_token_id": getattr(tokenizer, "pad_token_id", None) or getattr(tokenizer, "eos_token_id", None),
+        }
+
+        # Generate only with teacher (which has context) - this is the "ground truth" generation.
+        # The student will be trained to match this generation without seeing the context.
+        teacher_out = model.generate(teacher_prompt, attention_mask=torch.ones_like(teacher_prompt), **gen_kwargs)
```

Comment:
I tried this with the arxiv example (both `python examples/cartridge_self_study/arxiv_synthesize.py` and `python examples/cartridge_self_study/synthesize.py --model Qwen/Qwen2.5-0.5B-Instruct --corpus_path examples/cartridge_self_study/data/cartridges.tex --out_jsonl distill.jsonl --num_samples 1024`) and got an error here because `teacher_prompt` was a `transformers.tokenization_utils_base.BatchEncoding`.